common/src/heuristics.rs

```diff
@@ -8,53 +12,241 @@ const SIGSEGV: u32 = 11;
+///
+/// These heuristics must not be applied to frame #0 on any thread. The first frame of all threads
+/// contains the actual register state of that thread at the time that it crashed (if it's the
+/// crashing thread), or at the time it was suspended (if it is a non-crashing thread). These
+/// heuristics should only be applied to frames *after* frame #0 – that is, starting with frame #1.
+///
+/// Additionally, these heuristics assume that your symbolication implementation correctly handles
+/// addresses that occur within an instruction, rather than directly at the start of a valid
+/// instruction. This should be the case for any reasonable implementation, but is something to be
+/// aware of when deploying these changes.
+///
+/// ## x86 and x86-64
+///
+/// x86 uses variable-width instruction encodings; subtract one byte from the return address to
+/// derive an address that should be within the calling instruction. This will provide an address
+/// within a calling instruction found directly prior to the return address.
+///
+/// ## ARMv6 and ARMv7
 ///
-/// See <https://goo.gl/g17EAn> for detailed information on this topic.
+/// - **Step 1:** Strip the low order thumb bit from the return address. ARM uses the low bit to
+///   inform the processor that it should enter thumb mode when jumping to the return address. Since
+///   all instructions are at least 2 byte aligned, an actual instruction address will never have
+///   the low bit set.
+///
+/// - **Step 2:** Subtract 2 Bytes. 32-bit ARM instructions are either 2 or 4 bytes long, depending
+///   on the use of thumb. This will place the symbolication address within the likely calling
+///   instruction. All ARM64 instructions are 4 bytes long; subtract 4 bytes from the return address
+///   to derive the likely address of the calling instruction.
+///
+/// # More Information
+///
+/// The above information was taken and slightly updated from the now-gone *PLCrashReporter Wiki*.
+/// An old copy can still be found in the [internet archive].
+///
+/// [internet archive]:
+/// https://web.archive.org/web/20161012225323/https://opensource.plausible.coop/wiki/display/PLCR/Automated+Crash+Report+Analysis
+///
+/// [`caller_address`]: struct.InstructionInfo.html#method.caller_address
+#[derive(Clone, Debug)]
 pub struct InstructionInfo {
-    /// The address of the instruction we want to use as a base.
+    #[doc(hidden)]
+    #[deprecated = "Use InstructionInfo::new() instead"]
     pub addr: u64,
-    /// The architecture we are dealing with.
+    #[doc(hidden)]
+    #[deprecated = "Use InstructionInfo::new() instead"]
     pub arch: Arch,
-    /// This is true if the frame is the cause of the crash.
+    #[doc(hidden)]
+    #[deprecated = "Use is_crashing_frame() instead"]
     pub crashing_frame: bool,
-    /// If a signal is known that triggers the crash, it can be stored here.
+    #[doc(hidden)]
+    #[deprecated = "Use signal() instead"]
```

Comment:
Do we need accessors for this? If so, `set_signal` might be better.

Reply:
I'd say we probably don't, since the only use of this is to get the caller address and not to store this information.